src/main/scala/mesosphere/marathon/api/v2/AppsResource.scala

```diff
@@ -300,11 +306,38 @@ class AppsResource @Inject() (
   private def maybePostEvent(req: HttpServletRequest, app: AppDefinition) =
     eventBus.publish(ApiPostEvent(req.getRemoteAddr, req.getRequestURI, app))
 
-  private def search(cmd: String, id: String): Iterable[AppDefinition] = {
+  type LabelMatcher = (Map[String, String]) => Boolean
+
+  private def labelMatcher(labels: String): LabelMatcher = {
+    if (labels.indexOf(",,") != -1)
+      throw new InvalidQueryException(labels)
+
+    val pairs = for (pair <- labels.split(",")) yield pair.split("=")
+    val queryLabels = pairs.collect {
+      case Array(x, y, _*) => (x, y)
+    }.toMap
+
+    def matcher(appLabels: Map[String, String]): Boolean = {
+      val matches = for ((k, v) <- queryLabels) yield {
+        if (appLabels.contains(k) && appLabels(k) == v) {
```

Comment:
For this function, we only want to match if there are >1 true matches and 0 falses.  Because the result is OR'ed with appMatchesCmd or appMatchesId to determine whether we return this app, we want this to return false if there are zero queryLabels.